File: COTE/codewars/5-DirectionsReduction.js
//https://www.codewars.com/kata/550f22f4d758534c1100025a/solutions/javascript

function dirReduc(arr){
   while(true){
     let resetCheck=false;
     for(let i=1; i<arr.length;i++){
     let curIdx=i-1;
       if(isOp(arr[curIdx], arr[i])){
         arr=arr.slice(0,i-1).concat(arr.splice(i+1, arr.length))
         resetCheck=true;
         break
       }
     }
     if(!resetCheck){
       break
     }
   }
   return arr
 }
 
 function isOp(a,b){
   if((a=='SOUTH' && b=='NORTH') || (b=='SOUTH' && a=='NORTH') || (a=='WEST' && b=='EAST') || (a=='EAST' && b=='WEST')){
     return true
   }
   return false
 }

 
//s1
// use object & reduce from the back
function dirReduc(plan) {
   var opposite = {
     'NORTH': 'SOUTH', 'EAST': 'WEST', 'SOUTH': 'NORTH', 'WEST': 'EAST'};
   return plan.reduce(function(dirs, dir){
       if (dirs[dirs.length - 1] === opposite[dir])
         dirs.pop();
       else
         dirs.push(dir);
       return dirs;
     }, []);
 }

//trial1
function dirReduc(arr){
   const opposite = {
     "SOUTH":"NORTH", "EAST":"WEST", "NORTH":"SOUTH", "WEST":"EAST"
   }
   return arr.reduce((ac,cu)=>{
     if(opposite[cu]===ac[ac.length-1]){
       ac.pop();
     }else{
       ac.push(cu)
     }
     return ac;
   },[])
 }

//s2
// pattern.test(str), str.replace(pattern,''), str.match(/~/) ||[]
function dirReduc(arr) {
   var str = arr.join(''), pattern = /NORTHSOUTH|EASTWEST|SOUTHNORTH|WESTEAST/;
   while (pattern.test(str)) str = str.replace(pattern,'');
   return str.match(/(NORTH|SOUTH|EAST|WEST)/g)||[];
 }

File: COTE/codewars/5-DirectionsReduction.py
opposite={"SOUTH":"NORTH","NORTH":"SOUTH","EAST":"WEST", "WEST":"EAST"}
def dirReduc(arr):
    result=[]
    while(True):
        startOver=False
        for idx,el in enumerate(arr):
            next = arr[idx+1] if len(arr)-1>=idx+1 else ''
            if opCheck(el, next):
                arr=arr[0:idx]+arr[idx+2:]
                startOver=True
                break;
        if not startOver:
            break;
    return arr
def opCheck(a,b):
    if opposite[a]==b:
        return True
    return False


#1✨
opposite = {'NORTH': 'SOUTH', 'EAST': 'WEST', 'SOUTH': 'NORTH', 'WEST': 'EAST'}

def dirReduc(plan):
    new_plan = []
    for d in plan:
        if new_plan and new_plan[-1] == opposite[d]:
            new_plan.pop()
        else:
            new_plan.append(d)
    return new_plan


#2
# " ".join(arr)
# .replace
# .split()
# reculsive
def dirReduc(arr):
    dir = " ".join(arr)
    dir2 = dir.replace("NORTH SOUTH",'').replace("SOUTH NORTH",'').replace("EAST WEST",'').replace("WEST EAST",'')
    dir3 = dir2.split()
    return dirReduc(dir3) if len(dir3) < len(arr) else dir3